### PR TITLE
refactor(robot-server): consolidate and simplify error response logic

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR is a bit of an RFC, and comes out of a rabbit hole I fell into while trying to build some new endpoints.

In a lot of discussions we've had about error responses from our API, there's been a desire to have a concrete list of all errors that may be returned by the HTTP API. Ideally, this is something we can bake right into our OpenAPI Spec, but there were a couple things in the way of that effort:

- Our error models did not require the `id` field
- It was difficult to spec a specific error return from a given endpoint
- Returning an error from an endpoint was a little hard to read (my opinion)

This PR reorganizes and simplifies our models to allow for this sort of usage:

```py
from fastapi import APIRouter, status
from typing_extensions import Literal
from robot_server.errors import ErrorResponse, ErrorDetails


class SpecificProblem(ErrorDetails):
    id: Literal["SpecificProblem"] = "SpecificProblem"
    title: str = "Specific Problem"


class SomeModel(BaseModel):
    id: str
    name: str


router = APIRouter()


@router.get(
    path="/some-model",
    response_model=SomeModel,
    responses={
        status.HTTP_400_BAD_REQUEST: {"model": ErrorResponse[SpecificProblem]},
    },
)
def get_some_model():
    """Get some model."""
    # ...
    raise SpecificProblem(detail="A specific problem occurred").as_error(
        status.HTTP_400_BAD_REQUEST
    )
```

If we successfully set `responses` for all of our endpoints in this way, we will have an exhaustive list of every specific error response returned baked right into the OpenAPI Spec, without (necessarily) needing to maintain some giant, central `ErrorId` enum or similar.

## Changelog

Some top-of-line changes:

- Moved as much as possible related to error responses to `robot_server.errors`
- Removed usage of `V1HandlerError` in favor of `robot_server.errors.LegacyErrorResponse`
- Simplified the handler interface for returning an error response to flow directly from the response models themselves
- Genericized `ErrorResponse` so JSON schema knows what specific error is coming back
    - Removed redundant `status` field from `ErrorDetails`
    - Moved `links` field from `ErrorDetails` to `ErrorResponse` for parity with our regular responses
        - Good response: `{ "data": { ... }, "links": { ... } }`, error response: `{ "errors": [ ... ], "links": { ... } }`
- Fixed up a few things about our exception handlers
    - Legacy endpoints now return legacy errors during unhandled exceptions
    - Fixed our usage of the `source` field in validation errors to treat body fields, parameters, and headers differently

## Review requests

Going commit by commit might help on this one, since it's kind of big! Otherwise:

- Look at the code example above. Does this look like a good interface?
- Look at the code in the PR. Do changes make sense?
- Look at the tests added in the PR. Are they sufficient?

### Style guide proposals

I'd like this work to be an input into our API styleguide. Would love to hear thoughts, opinions, ideas, and disagreements as you read through this!

So far, I have:

- Only route handlers should know about status codes and response bodies
    - Anything below a route handler should not be throwing `ApiError` exceptions
    - Instead, they should throw something specific to their own logic and have the router catch and respond accordingly
- Anything else?

## Risk assessment

Low to medium. I removed some unused fields from error responses and tweaked things here and there, but I don't think they'll have any effect on clients / the app.

Integration tests are still passing, but I did have to make some changes to some of them, so please **inspect Tavern YAML changes carefully**.

### CODE REVIEW

**Comment by SyntaxColoring** [comment]
Specifying errors exhaustively endpoint-to-endpoint would be great. 🎉 

But what happens if an endpoint has multiple error models for the same HTTP status code? For example, maybe there are two ways a request can be `403 FORBIDDEN`, and we want to provide different structured details for each way. Then, we're right back where we started, I think: we have to lump the models together in a union.

I guess this is the best that FastAPI can do?

-----

For other reviewers: the [Additional Responses](https://fastapi.tiangolo.com/advanced/additional-responses/) section is the relevant part of FastAPI's very verbose and unhelpfully-organized documentation. 🙃

File: robot-server/robot_server/errors/exception_handlers.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                return V1_TAG in route.tags
+
+    return False
+
+
+def _format_validation_source(
+    parts: Sequence[Union[str, int]]
+) -> Optional[ErrorSource]:
+    """Format a validation location from FastAPI into an ErrorSource."""
+    if parts[0] == "body":
```
Comment (amitlissack) [inline robot-server/robot_server/errors/exception_handlers.py]: What happens if this function raises?

**Review by amitlissack — COMMENTED** [review]
My impression is that this is a vastly huge improvement on an area of the robot-server that got very little attention. 

I really like the separating out of "error stuff"  and the resulting slimmer `app.py`.

👍

File: robot-server/robot_server/errors/exception_handlers.py (hunk 1 above)
Comment (mcous) [reply robot-server/robot_server/errors/exception_handlers.py]: Oof, bad things. What are you thinking for increased safety? Check bounds before array access and/or a `try ... except IndexError`?

Comment (amitlissack) [reply robot-server/robot_server/errors/exception_handlers.py]: Either of those would be fine really.

File: robot-server/robot_server/errors/exception_handlers.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+async def handle_validation_error(
+    request: Request,
+    error: RequestValidationError,
+) -> JSONResponse:
+    """Map a validation error from the framework to an API response."""
+    validation_errors = error.errors()
+
+    if _route_is_legacy(request):
+        message = "; ".join(
+            [
```
Comment (amitlissack) [inline robot-server/robot_server/errors/exception_handlers.py]: You don't need the `[` and `]`. A generator expression works here.

**Review by amitlissack — APPROVED** [review]
Looks good.

**Comment by mcous** [comment]
For this specific question:

> But what happens if an endpoint has multiple error models for the same HTTP status code? For example, maybe there are two ways a request can be 403 FORBIDDEN, and we want to provide different structured details for each way

This ends up producing a pretty nice OpenAPI spec / docs:

```py
@router.get(
    path="/some-model",
    response_model=ResponseModel[SomeResource],
    responses={
        status.HTTP_400_BAD_REQUEST: {
            "model": Union[
                ErrorResponse[SpecificProblem],
                ErrorResponse[OtherProblem],
            ],
        },
    },
)
```

**Review by sanni-t — APPROVED** [review]

**Pull request merged**